### ISSUE TITLE
Odd column ordering in data output, based on non response for first response.

### ISSUE THREAD

**Issue by frankgasking** [psychopy/psychopy#1097]
Hi all,

A bit of an odd issue today - but we've found that if you have a routine with a response object, and you don't respond for the very first trial - the .RT column for that response object will not appear in sequence alongside the accuracy and keys columns for the object - it appears elsewhere as one of the columns.If you respond for the very first trial, it does appear in the same object sequence as you would expect.

So it seems that when RT is blank (for no response), something in the datafile output causes the sequence to change.  This causes some confusion when people are just doing straightforward merges of their CSV files - as the columns don't fully tie up.  

Attached is a very quick dummy script to demonstrate - and the two data files will show the differences as described above.   See how for Participant 2, key_resp_2.rt, suddenly jumps so it follows after key_resp_3.rt.
[test-script.zip](https://github.com/psychopy/psychopy/files/105057/test-script.zip)

We are currently fixed on version 1.82.01 for this year, due to upgrade in July in preparation for the new University year.

Many thanks,

Frank

**Comment by jeremygray** [comment]
Other people have also noticed this issue (variable output order). You might reload the .psydat file, and save the data again. The newly saved data might be in the typical column order, but I am not sure about that.

In principle, I think this problem can and should be handled during post-processing, not by PsychoPy upon data generation. Doing the "right thing" in all situations at experiment run-time, and without it being a hassle for the experimenter, would be quite a lot of effort. And I think its best dealt with during post-processing, where it is more natural and would be relatively easy to do. Using pandas, data columns are addressable by name (not solely by position or numerical index).

It would be straightforward for PsychoPy to sort columns alphabetically prior to writing out data files. This could help provide uniformity, but there could still be some failures (due to missing columns, if there's never an RT for a given type of trial). If alphabetical column-name output would be useful (even if not solving all possible issues), I could see adding that as an option.

**Comment by frankgasking** [comment]
Thanks Jeremy.    Longer term, it would be useful feature to have something to help those with no programming background and to avoid confusion prior to doing merges - even as a separate post-processing tool.

I'm going to build a simple batch script in the meantime to handle any cases that we get for the above, and then carry out a merge if all is good.  I'll make that available on our public github, in case its useful for others.

**Comment by jeremygray** [comment]
@qbilius Jonas, do you have someting in psychopy_ext to handle combining multiple data files in which the column orders may differ?

**Comment by qbilius** [comment]
Are you guys talking about storing data in the TrialHandler? I don't use it as I always found it more confusing than helpful. So there aren't any functions for this purpose in psychopy_ext.

However, in order to merge data files, I think pandas might offer a solution. I'm pretty sure when merging multiple DataFrames in pandas, column order is automatically adjusted.

**Comment by peircej** [comment]
Haha, fair call (in my defence, when that was written the Python community was wondering whether to use Numeric or numarray, before numpy and many years before pandas). The new TrialHandler2 is much much cleaner.
I think the discussion here, though, is actually about the ExperimentHandler outputs. Those are going to be trickier to handle in an 'organised' way. We could just sort alphabetically and we'll always get the same order but it won't be sensible. The current way is using the order that the variables are created (so practice trial variables naturally come before main trial variables).

Basically, it will take some thought to do this right (track where variables come from and alphabetise within a group of variables?)

**Comment by jeremygray** [comment]
thanks @qbilius! yes, merging files using pandas will do the right thing using pandas.concat(). definitely seems like the way to go to me.

**Comment by jeremygray** [comment]
This will combine all data files into a single pandas dataframe `df`, properly aligning columns having the same header name, regardless of the order within the files:

```python:
import glob
import pandas as pd
files = glob.glob('*.csv')  # find all PsychoPy data files, csv format
df = pd.read_csv(files.pop())
for f in files:
    df = pd.concat([df, pd.read_csv(f)])
```

**Comment by jeremygray** [comment]
Or simpler:

```python
import glob
import pandas as pd
files = glob.glob('*.csv')
if files:
    df = pd.concat([pd.read_csv(f) for f in files])
```

**Comment by frankgasking** [comment]
Thanks everyone for the great responses - including the pandas solution for
merging.  It's one of the many reasons I love Python! :)

On Sun, Jan 31, 2016 at 2:52 PM, Jeremy R. Gray <notifications@github.com>
wrote:

> Or simpler:
>
> import globimport pandas as pd
> files = glob.glob('*.csv')if files:
>     df = pd.concat([pd.read_csv(f) for f in files])
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/psychopy/psychopy/issues/1097#issuecomment-177515226>.
>



-- 

Frank

----------------------------------------
Visit my retro blog at: http://fgasking.wordpress.com/

Also the GTW project -
http://www.gamesthatwerent.com (main)
http://www.gtw64.co.uk (c64)

**Comment by jeremygray** [comment]
Seems resolved.
